central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,646 @@
+}
+
+func (s *scheduler) SubmitReport(request *ReportRequest, reSubmission bool) (string, error) {
+	request.Ctx = selectContext(request)
+	var err error
+	if request.ReportMetadata.GetReportStatus().GetReportRequestType() == storage.ReportStatus_ON_DEMAND {
+		userHasAnotherReport, err := s.doesUserHavePendingReport(request.Ctx, request.ReportConfig.GetId(), request.ReportMetadata.GetRequester().GetId())
+		if err != nil {
+			return "", err
+		}
+		if userHasAnotherReport {
+			return "", errors.Errorf("User already has a running report for config ID %s", request.ReportConfig.GetId())
+		}
+	}
+
+	request.ReportMetadata.ReportStatus.RunState = storage.ReportStatus_WAITING
+	request.ReportMetadata.ReportStatus.QueuedAt = types.TimestampNow()
+	if !reSubmission {
+		request.ReportMetadata.ReportId, err = s.reportMetadataStore.AddReportMetadata(request.Ctx, request.ReportMetadata)
+	} else {
+		err = s.reportMetadataStore.UpdateReportMetadata(request.Ctx, request.ReportMetadata)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	s.lock.Lock()
+	s.queuedReports = append(s.queuedReports, request)
+	s.lock.Unlock()
+	s.sendNewReports <- true
+
+	return request.ReportMetadata.GetReportId(), nil
+}
+
+func (s *scheduler) doesUserHavePendingReport(ctx context.Context, configID string, userID string) (bool, error) {
+	query := search.NewQueryBuilder().
+		AddExactMatches(search.ReportConfigID, configID).
+		AddExactMatches(search.ReportState, storage.ReportStatus_WAITING.String(), storage.ReportStatus_PREPARING.String()).
+		AddExactMatches(search.ReportRequestType, storage.ReportStatus_ON_DEMAND.String()).
+		ProtoQuery()
+	runningReports, err := s.reportMetadataStore.SearchReportMetadatas(ctx, query)
+	if err != nil {
+		return false, err
+	}
+	for _, rep := range runningReports {
+		if rep.GetRequester().GetId() == userID {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (s *scheduler) CancelReport(reportID string) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (s *scheduler) runReports() {
+	defer s.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case req := <-s.reportsToRun:
+			log.Infof("Executing report '%s' at %v", req.ReportConfig.GetName(), time.Now().Format(time.RFC822))
+			go s.sendReportResults(req)
+		}
+	}
+}
+
+func (s *scheduler) scheduleReports() {
+	maxConcurrency := int32(env.ReportExecutionMaxConcurrency.IntegerSetting())
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case <-s.sendNewReports:
+			numRoutines := s.numRoutines.Load()
+			// Checking len(s.reportsToRun) < maxConcurrency (which is also the capacity of buffered channel reportsToRun) here
+			// will ensure that we never attempt to write to full channel reportsToRun. This will further ensure that
+			// writes to reportsToRun in selectRunnableReports() are never blocked.
+			if int32(len(s.reportsToRun)) < maxConcurrency && numRoutines < maxConcurrency {
```

Comment:
`numRoutines` can be incorrect by the time you get to here.

Reply:
Changed to using weightedSemaphore instead of atomic integer for limiting concurrent routines.